### PR TITLE
Add proof diffs to Coqtail

### PR DESCRIPTION
**Pull request by j-hui**
Adds proof diffs to Coqtail (#57).

Summary of changes on Python side:
- Replace the `_to_string()` function with `_to_richpp()`, which gives a list of tagged tokens for Coq 8.6+
- Maintain addition `goal_hls` state to keep track of all the highlights to be applied to the goal panel
- `pp_goals()` now builds and returns a list of lines and highlights (it was previously returning a single text string, which was later being `.split(\n)` prior to be being set to `goal_msg`)
- Send both lines and highlights to Vim side (send empty list of highlights for the info panel)

Summary of changes on Vim side:
- Added some new highlight groups for diffs
- Add buffer state to keep track of highlighted ranges
- Receive additional `richpp` information from Python side
- Use `richpp` to highlight proof diffs

Remaining items:
- [x] Test on Vim, Gvim, Neovim, etc.
- [x] Test with different Coq versions
- [x] Preserve newlines in `line_and_highlights()` function
- [x] Better default highlight colors
- [x] ~Support turning diffs on and off from Coqtail (rather than having to use `Set Diffs "on".`)~ (Turns out `:Coq Set Diffs "on"` is already supported)

### CODE REVIEW

**Comment by hmontero1205** [comment]
following

**Comment by j-hui** [comment]
According to CI this feature is doing quite poorly with Python2.7... Is this something I should fix and support?

Then again, I have done very little testing so far.

**Comment by j-hui** [comment]
Oh, I should also note that I encountered some ambiguity when it came to the way `start.` and `end.` tags were used. I expected them to only be used as self-closing tags, but I encountered messages that looked like the following:

```xml
 <pp> <start.diff.added.bg/> c : <start.diff.added> <constr.variable> B </constr.variable> </start.diff.added> <end.diff.added> <constr.variable> x </constr.variable> </end.diff.added> <end.diff.added.bg/> </pp>
```

I am asking for clarification [here](https://github.com/coq/coq/issues/13668).

**Comment by j-hui** [comment]
Turns out `:Coq Set Diffs "on"` already works.

File: syntax/coq.vim
Comment (whonore) [inline syntax/coq.vim:46]: These should reuse the default Vim diff highlighting groups (`DiffAdd`, `DiffChange`, `DiffDelete`, `DiffText`). For example, something like `hi def link CoqtailDiffAdded DiffAdd`.

File: python/xmlInterface.py
Comment (whonore) [inline python/xmlInterface.py:43]: For this to work in Python 2 you also need to mock all the types that you pass as the first argument of `cast`. I think a simpler option is to not use `cast` and instead just use `# type: ignore` comments where necessary with a comment stating the expected type. Then when #159 eventually happens we can use `cast` again.

**Comment by whonore** [comment]
Thanks, this looks like a really good start. I left a couple quick comments, and I'll try to do a more thorough review soon.

> this feature is doing quite poorly with Python2.7 

I would like to get a version of this that supports Python 2, but then I'm hoping to do #159 soon after. Anywhere you have to make a decision solely to cater to Python 2, just leave a comment so we remember to clean it up when we switch to Python 3 only.

> Support turning diffs on and off from Coqtail

Even though people can do `:Coq Set Diffs "on"` it might be nice to provide a Vim option (e.g., `g:coqtail_diffs = "always"`) to enable them automatically. Maybe that belongs in a later PR though.

**Comment by whonore** [comment]
Also, I'm running pretty low on credits for Travis builds so depending on how often you push you might find that the CI stops running. It's not a big deal, I'm in the process of migrating to GitHub Actions (#167), which is proving to be trickier than I thought, but it's almost all working. I'll let you know when I merge it and then you should rebase on top of it.

File: syntax/coq.vim
Comment (j-hui) [inline syntax/coq.vim:45]: Do we want to give users the option to specify their own proof diff highlighting, e.g., using `g:CoqtailHighlight`?

**Comment by j-hui** [comment]
> it might be nice to provide a Vim option (e.g., `g:coqtail_diffs = "always"`)

This is probably preferable. One thing that's kind of weird is that the setting is stateful, and tied to the state of the proof progress. For example:

```coq
Example momopopo (P Q R: Prop): P -> (P -> Q) -> (Q -> R) -> R.
Proof.
  intro p.
  intro pq.
  intro qr. (* :Coq Set Diffs "on" *)
  apply pq in p.
  apply qr.
  apply p.
Qed.
```
If I turn on diffs halfway through the proof (where it says in the comments), all the following tactics will show diffs, including if I rewind. But If I rewind past the point where I called `:Coq Set Diffs "on"`, diffs stop appearing, and if I step forward again, they will also be gone. So nothing terrible (you can just turn it back on again) but it's just a very weird behavior.

Perhaps the default behavior should be to turn it on at the beginning of any Coqtail instance, and then allow the user to toggle the highlighting on and off from the Vim side.

File: syntax/coq.vim
Comment (whonore) [reply syntax/coq.vim:45]: If I understand `:h :hi-default` correctly users should be able to override these default settings by putting something like `hi CoqtailDiffAdded ctermfg=1` in their .vimrc or anything that gets sourced before Coqtail's syntax file. For that matter, the same should be true of `CoqtailChecked` and `CoqtailSent` so `g:CoqtailHighlight` shouldn't be necessary at all.

Comment (j-hui) [reply syntax/coq.vim:45]: I see, so I'll change mine here to use that instead.

Comment (j-hui) [reply syntax/coq.vim:45]: I tried removing the `g:CoqtailHighlight` from my `.vimrc` (which I used for custom `CoqtailChecked` and `CoqtailSent` highlighting) and it stopped working.. I'm not totally sure why. I'll need to investigate that further but I think we'll want to keep `g:CoqtailHightligh` here for now, but also keep this PR as it is like this?

Comment (whonore) [reply syntax/coq.vim:45]: Hm, if I put `hi link CoqtailChecked Error` in my .vimrc (and no `g:CoqtailHighlight`) it shows up as red as expected. In any case, I agree we should leave `g:CoqtailHighlight` for now and go through some sort of deprecation process in a later PR so people aren't surprised by their custom highlighting suddenly not working.

**Comment by j-hui** [comment]
> Even though people can do :Coq Set Diffs "on" it might be nice to provide a Vim option (e.g., g:coqtail_diffs = "always") to enable them automatically. Maybe that belongs in a later PR though.

I have something like this ready to push. Should I include it in this PR or save it for another?

**Comment by j-hui** [comment]
I'm not sure why https://github.com/whonore/Coqtail/runs/1619603883 is taking so long... in any case, it looks like I'm failing the `test_dispatch_unicode` test case in Python2 because it no longer recognizes the unicode string as an instance of `str`.

Unfortunately, the unit test seems a little difficult for me to replicate locally. When I install Python2, `tox` still seems to pass:

```
...
____ summary ____
  unit-py27: commands succeeded
ERROR:  unit-py36: InterpreterNotFound: python3.6
  coq-py27: commands succeeded
ERROR:  coq-py36: InterpreterNotFound: python3.6
  vim: commands succeeded
```

**Comment by j-hui** [comment]
I'm not sure what's making that CI build take so long.. but in any case, I got tests running on my local environment, and was able to fix issues with python2.

Basically, the issue was that I was checking `not isinstance(_, Text)` to distinguish between plain strings and richpp strings (which would come back as a list of string/tag pairs), except `Text` isn't imported in python2. I initially switched this to `not isinstance(_, str)`, but this didn't work in Python2 when encountering unicode. So the fix was to just do `isinstance(_, list)`.

**Comment by whonore** [comment]
> I have something like this ready to push. Should I include it in this PR or save it for another?

You can go ahead and include it here.

> I'm not sure what's making that CI build take so long/what has it blocked

I'm out of free credits on Travis. I'm going to do my best to get #167 working today.

> but this didn't work in Python2 when encountering unicode

Your fix should be fine, but another option is to use `six.text_type` which is `str` for Python 3 and `unicode` for Python 2. There's also `six.string_types` which is a superclass of both `str` and `unicode` in Python 2 and `six.binary_type` which is `bytes` in 3 and `str` in 2.

**Comment by j-hui** [comment]
> I'm out of free credits on Travis.

Uhhh if that was due to my pushes... I'm really sorry. But maybe that's why it's hanging?

> another option is to use `six`

This seems like the more responsible/stable thing to do. I'll do that.

**Comment by whonore** [comment]
> Uhhh if that was due to my pushes... 

It's no problem, it was going to happen soon anyway and once I finish switching to GitHub Actions I won't need them anymore.

File: autoload/coqtail/panels.vim
Comment (whonore) [inline autoload/coqtail/panels.vim:250]: Same here.

File: syntax/coq.vim
Comment (whonore) [inline syntax/coq.vim:45]: Should one of these not use `DiffAdd` (`:h hl-DiffAdd` says it's for an added line)? I'm having trouble finding documentation on what each of the Coq diff types means so it's possible these are correct as-is.

**Review by whonore — CHANGES_REQUESTED** [review]
I only did a pass on the code, I haven't actually tested the functionality thoroughly yet. Also please run `tox -e format` to run `black` and `isort`.

File: autoload/coqtail.vim
Comment (j-hui) [reply autoload/coqtail.vim:409]: Another issue I just thought of here is that this will not work for versions of Coq before proof diffs were available.

How should this version checking happen on the vim side? Or should it be delegated to the Python side?

Comment (j-hui) [reply autoload/coqtail.vim:409]: Looks like proof diffs are supported [since version 8.9.0](https://coq.github.io/doc/master/refman/changes.html#id744).

I adapted your code that checks for Coq version support, to determine when to query with `Set Diffs "{arg}"` at start (called in `coqtail#start()`):

```
function! s:setup_proofdiffs(coqtail_version) abort
  let l:proof_diffs_arg = coqtail#util#getvar([b:, g:], 'coqtail_auto_enable_proof_diffs', "")
  if l:proof_diffs_arg == ""
    return
  endif

  let l:versions = map(split(a:coqtail_version, '\v(\.|pl)'), 'str2nr(v:val)')
  " Pad missing version numbers with 0
  while len(l:versions) < 3
    let l:versions = add(l:versions, 0)
  endwhile

  for l:supp in s:proof_diffs_supported
    let l:is_sup = 1
    for l:idx in range(3)
      if l:supp[l:idx] != l:versions[l:idx] && l:supp[l:idx] != -1
        let l:is_sup = 0
        break
      endif
    endfor
    if l:is_sup
      call s:call("query", "", 0, {"args": ["Set", "Diffs", '"' . l:proof_diffs_arg . '"' ]})
      return
    endif
  endfor
endfunction
```

Is there a better way to do this? Seems a bit clunky for just one feature..

File: syntax/coq.vim
Comment (j-hui) [reply syntax/coq.vim:45]: `DiffAdd` is for highlighting the entirety of a new line, which Coqtop doesn't have a separate kind of highlighting for (it just shows the entire line as `CoqtailDiffAdded`).

I'm using `DiffChange` and `DiffText` because those are respectively meant for the background and content of a changed line, and are thus likely to be defined to work together. I avoided using `DiffChange` and `DiffAdd` because the colors are more likely to clash.

File: autoload/coqtail.vim
Comment (j-hui) [reply autoload/coqtail.vim:409]: Also, for what it's worth, I tried using the `"removed"` option. I'm not really sure what Coqtop is trying to do based on what Coqtail is showing, but it makes it really hard to tell what's going on.

Moving between a context of:

```
1 subgoal
C : Prop
T : Set
α : T -> Prop
H : exists x : T, C -> α x
c : C
========================= (1 / 1)
exists x : T, α x
```

and 

```
1 subgoal
C : Prop
T : Set
α : T -> Prop
x : T
H : C -> α x
c : C
========================= (1 / 1)
exists x0 : T, α x0
```

I get this:

image

File: syntax/coq.vim
Comment (j-hui) [reply syntax/coq.vim:45]: Actually, `CoqtailDiffRemovedBg` should also be `DiffDelete`. Otherwise it's impossible to tell what's going on..

Comment (j-hui) [reply syntax/coq.vim]: See [comment above](https://github.com/whonore/Coqtail/pull/169#discussion_r550818445) for what that looks like now.

File: autoload/coqtail.vim
Comment (whonore) [reply autoload/coqtail.vim:409]: I just merged #178, which adds a `coqtail#version#atleast` function that you can use to check if `b:coqtail_version` is 8.9.0 or greater.

**Comment by j-hui** [comment]
ok @whonore I think I addressed all of your requested changes. Let me know if there's anything else!

**Comment by whonore** [comment]
Thanks, sorry for the delay. I’ll test it out this weekend. If I find any problems that seem easy to fix I may just do it myself, if that’s ok with you. Otherwise I’ll do another review and hopefully we can get this merged.

**Comment by j-hui** [comment]
Yeah no worries, I just didn't want to hold up #159. And please feel free to fix whatever (:

**Comment by whonore** [comment]
Ok, I think it looks good now. I want to test it out for a couple days this week and if I don't run into any issues I'll merge it.

**Review by whonore — APPROVED** [review]

**Comment by whonore** [comment]
I ran into what seems to be a bug where some whitespace is lost when proof diffs are on. The following example demonstrates what I mean.

```coq
From Coq Require Import List.
Axiom set : Type.
Axiom card : set -> nat.
Axiom union : set -> set -> set.
Axiom intersect : set -> set -> set.
Axiom el : set.
Axiom ss : list set.

Definition x := fold_left union ss (union el el).
Definition y := fold_left Nat.add (map card ss) (card el).
Definition z := fold_left intersect ss (intersect el el).

Goal card x = y - card z.
Proof.
  unfold x, y, z.

  (*
card (fold_left union ss (union el el)) =fold_left Nat.add (map card ss) (card el) -
card (fold_left intersect ss (intersect el el))
   *)

  Set Diffs "off".

  (*
card (fold_left union ss (union el el)) =
fold_left Nat.add (map card ss) (card el) -
card (fold_left intersect ss (intersect el el))
   *)

Admitted.
```

File: python/coqtail.py
Comment (j-hui) [inline python/coqtail.py:68]: Yeesh, good catch. Thanks for fixing that!

**Comment by whonore** [comment]
@j-hui thanks for all your work on this. It's a cool feature that definitely was not trivial to add, but Coqtail is now much better for supporting it.

**Pull request merged**
